### PR TITLE
Implement _type selection dropdown (TACOSTRAT-1516)

### PR DESCRIPTION
**Pull request by cgolemme**
# Summary

This PR adds a dropdown for selecting _type parameters for the bulk-export request and displays the bulk export request string to the user at the top of the page. [JIRA Issue](https://jira.mitre.org/browse/TACOSTRAT-1516)

Also removes the patient dropdown selection for ids because bulk-export-server does not support $export request on a patient id, like it does with group ids.

## New behavior

The /query-builder page now has tabs for each of the selection types. The type selection tab has a searchable dropdown menu to select types to add to the query. The query now appears at the top of the page in an input box. When you add types to the query, you can see the string change at the top. The button on the right will lead to the next page in the future, for now it does nothing. There is a copy button next to the input at the top, which was mainly just for my own sake in testing that the export request string was built correctly. I left it in because it could be a good functionality to have (I found I used it a lot).

## Code changes
- `data/supportedResources.ts` - pulled from the bulk export server it contains the supported resources that a user can choose to query from.
- `next.config.mjs` - added reactStrictMode: false to prevent error when running npm run dev that caused double rendering issues. It doesn't effect npm run build and npm run start.
- `package-lock.json`
- `package.json` -- added Tabler icons and recoil
- `src/app/query-builder/layout.module.css` some styling for layout component.
- `src/app/query-builder/layout.tsx` created query builder layout with header, main and footer sections.
- `src/app/query-builder/page.tsx` created the query builder page to implement the above functionalities.
- `src/app/query-builder/providers.tsx` component to use recoil and appshell without causing issues with the layout file.
- `src/app/query-selector/layout.tsx` changed to mirror changes on the builder page
- `src/app/query-selector/page.module.css` changed styling to mirror changes on builder page
- `src/components/type-params-page.tsx` created the types param page with the dropdown to select types
- `src/components/query-string.tsx` created to display the bulk-export request string as the user selects parameters
- `src/state/type-params-state.ts` recoil state for _type parameters
- `src/util/exportRequestBuilders.ts` contains functions to build the export request string.
- `src/components/query-selector/build-query-button.tsx` changed the name of the query string parameter type to exportType to be more specific.

## Testing
- `npm run check`
- `npm run build`
- `npm run start`
- Make sure functionality described above and in TACOSTRAT-1516 is implemented. Make sure you can select types, remove them, remove all of them, and make sure that the query string at the top reflects that change.

### CODE REVIEW

File: src/app/query-builder/layout.tsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { AppShellFooter, AppShellHeader, AppShellMain, Box, Title } from '@mantine/core';
+import Link from 'next/link';
+import Providers from './providers';
+
+import classes from './layout.module.css';
+
+export default function BuilderLayout({ children }: { children: React.ReactNode }) {
```
Comment (elsaperelli) [inline src/app/query-builder/layout.tsx:8]: Question: is there a reason that this and `providers.tsx` are separate? It seems to me like they are both kind of small amounts of code I don't think either is reused so they could just go in one?

File: src/app/query-builder/layout.tsx [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  return (
+    <Providers>
+      <AppShellHeader>
+        <Box component={Link} href="/query-selector" className={classes.appHeaderTitle}>
+          <Title>Bulk-export-app</Title>
+        </Box>
+      </AppShellHeader>
+      <AppShellMain>{children}</AppShellMain>
+      <AppShellFooter></AppShellFooter>
```
Comment (elsaperelli) [inline src/app/query-builder/layout.tsx]: Do we need the AppShellFooter? If we do, perhaps `<AppShellFooter />` :)

File: src/app/query-builder/page.tsx [hunk 3]
```diff
@@ -1,23 +1,61 @@
...
-      <QueryBuilderContent />
-    </Suspense>
-  );
+/*
+ * Properties of the query string that can be passed to this page
+ *
+ * exportType - type of export request from query-selector page
+ * id         - possibly undefined in export type is "system."
+ *              Otherwise, either group or patient id to query.
+ */
```
Comment (elsaperelli) [inline src/app/query-builder/page.tsx]:
```suggestion
 */
```

File: src/app/query-builder/page.tsx [hunk 4]
```diff
@@ -1,23 +1,61 @@
...
+export interface SearchParamsProps {
+  exportType: SupportedExportTypes;
+  id?: string;
 }
 
-function QueryBuilderContent() {
-  const searchParams = useSearchParams();
-  const type = searchParams.get('type');
-  const id = searchParams.get('id');
+const parametersHelpText = `Add parameters to your bulk export query`;
```
Comment (elsaperelli) [inline src/app/query-builder/page.tsx]: This is short enough and not reused (that I can see) so I would just put this in label no need to define

File: src/util/exportRequestBuilders.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+/**
+ * Builds bulk export request string from a list of parameters, a SupportedExportType, and an id.
+ * Note: currently only supports type parameters.
+ */
+export function buildExportRequestString(request: BuilderRequest) {
+  const { exportType, id, typeParams } = request;
+  const baseUrl = `${process.env.NEXT_PUBLIC_HOST}:${process.env.NEXT_PUBLIC_PORT}`;
+  const queryString = typeParams.length === 0 ? '' : `?_type=${typeParams.toString()}`;
```
Comment (elsaperelli) [inline src/util/exportRequestBuilders.ts:22]:
```suggestion
  const queryString = typeParams.length !== 0 ?? `?_type=${typeParams.toString()}`;
```
I think? Maybe test rq

**Review by elsaperelli — CHANGES_REQUESTED** [review]
Woo good stuff! Functionality looks great, one thing I noticed was that when I select a Patient and then go to the query builder, the Patient id does not seem to be included in the request string. It works for Group though, so hopefully a quick fix. Otherwise, just some small comments!

File: src/app/query-builder/layout.tsx (hunk 2 above)
Comment (cgolemme) [reply src/app/query-builder/layout.tsx]: Don't need it. I was messing around with putting some content in the footer but I didn't end up using it.

File: src/util/exportRequestBuilders.ts (hunk 5 above)
Comment (cgolemme) [reply src/util/exportRequestBuilders.ts:22]: eslint doesn't like this `Unexpected constant nullishness on the left-hand side of a ??  expression.eslint`[no-constant-binary-expression](https://eslint.org/docs/latest/rules/no-constant-binary-expression). I am not sure if there is a better way. This function will probably change when we have more parameters. I might be able to do something nicer then, like selectively call the function to add type parameters if they exist?

File: src/app/query-builder/layout.tsx (hunk 1 above)
Comment (cgolemme) [reply src/app/query-builder/layout.tsx:8]: I do believe it is fine to put in the layout. I am still confused by next. RecoilRoot needs to be in a client component with the `"use client";` directive at the top, and I thought that layouts cannot be client components. That's why i extracted it to another component. It seems that just Root Layouts cannot be client components? Anyways, this is what i was reading that brought me to that conclusion https://stackoverflow.com/questions/74992326/does-use-client-in-next-js-13-root-layout-make-whole-routes-client-component. It doesn't seem to be very well documented in the next documentation. Either way I removed the Providers component and put it in the layout.

File: src/util/exportRequestBuilders.ts (hunk 5 above)
Comment (elsaperelli) [reply src/util/exportRequestBuilders.ts:22]: Ah nah no worries then! Not worth fussing over, this works! Thanks for checking.

File: src/app/query-builder/layout.tsx (hunk 1 above)
Comment (elsaperelli) [reply src/app/query-builder/layout.tsx:8]: Ah gotcha yeah this is a newer version of Next than I am used too so thanks for the info! This works so all good.

**Review by elsaperelli — APPROVED** [review]
Woo lgtm!

**Pull request merged**